File: frontend/src/utils/pembayaranUtils.ts
import type { Kontrak } from '@/types'

/** Selisih pembulatan PPh (senilai backend Superman). */
export const PAYMENT_LUNAS_TOLERANCE = 1.5

export function isInvoicePaid(paid: number, invoiceTotal: number): boolean {
  return invoiceTotal > 0 && paid >= invoiceTotal - PAYMENT_LUNAS_TOLERANCE
}

/** PPh yang dipotong pembeli dari transfer bersih (net). */
export function pphOnNetTransfer(nominal: number, kontrak: Kontrak | null | undefined): number {
  if (!kontrak || nominal <= 0 || kontrak.is_pph !== 'true') return 0

  const ppnRate = (kontrak.ppn_persen || 0) / 100
  const pphRate = (kontrak.pph_persen || 0) / 100
  const hasPpn = kontrak.is_ppn !== 'false' && ppnRate > 0
  const denom = hasPpn ? 1 + ppnRate - pphRate : 1 - pphRate
  if (denom <= 0) return 0

  const pokok = nominal / denom
  return Math.round(pokok * pphRate)
}

/** Pelunasan untuk cek lunas/Superman — PPh dipotong pembeli selalu dihitung jika kontrak kena PPh. */
export function effectivePelunasan(
  nominal: number,
  _isPphDisetor: string | null | undefined,
  kontrak: Kontrak | null | undefined,
): number {
  const base = Number(nominal) || 0
  if (kontrak?.is_pph !== 'true') return base
  return base + pphOnNetTransfer(base, kontrak)
}

export function paymentProgressPercent(paid: number, invoiceTotal: number): number {
  if (invoiceTotal <= 0) return 0
  const pct = (paid / invoiceTotal) * 100
  if (isInvoicePaid(paid, invoiceTotal)) return 100
  return Math.min(99.9, Math.round(pct * 10) / 10)
}

File: services/pembayaran_utils.py
"""Perhitungan pelunasan pembayaran (termasuk PPh disetor pembeli)."""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    import models

# Selisih di bawah Rp 100 dianggap lunas (pembulatan PPh/transfer).
PAYMENT_LUNAS_TOLERANCE = 99


def is_invoice_paid(paid_total: float, invoice_total: float) -> bool:
    return invoice_total > 0 and paid_total >= invoice_total - PAYMENT_LUNAS_TOLERANCE


def pph_on_net_transfer(nominal: float, kontrak: "models.Kontrak | None") -> float:
    """PPh yang dipotong pembeli saat transfer bersih."""
    if not kontrak or nominal <= 0:
        return 0.0
    if str(getattr(kontrak, "is_pph", "false")).lower() != "true":
        return 0.0

    ppn_rate = float(getattr(kontrak, "ppn_persen", 0) or 0) / 100
    pph_rate = float(getattr(kontrak, "pph_persen", 0) or 0) / 100
    has_ppn = str(getattr(kontrak, "is_ppn", "true")).lower() != "false" and ppn_rate > 0
    denom = (1 + ppn_rate - pph_rate) if has_ppn else (1 - pph_rate)
    if denom <= 0:
        return 0.0

    pokok = nominal / denom
    return round(pokok * pph_rate)


def effective_pelunasan(
    nominal: float,
    is_pph_disetor: str | None,
    kontrak: "models.Kontrak | None",
) -> float:
    """Pelunasan untuk cek lunas/Superman.

    PPh yang dipotong pembeli selalu dianggap bagian pembayaran jika kontrak
    kena PPh. Parameter is_pph_disetor hanya untuk pelacakan di Laporan Digital.
    """
    _ = is_pph_disetor
    base = float(nominal or 0)
    if str(getattr(kontrak, "is_pph", "false") if kontrak else "false").lower() != "true":
        return base
    return base + pph_on_net_transfer(base, kontrak)


def pembayaran_paid_total(payments: list["models.Pembayaran"], kontrak: "models.Kontrak | None") -> float:
    return sum(
        effective_pelunasan(p.nominal_transfer, p.is_pph_disetor, kontrak)
        for p in payments
    )


def max_nominal_transfer(
    sisa_pelunasan: float,
    kontrak: "models.Kontrak | None",
) -> float:
    """Batas nominal transfer agar pelunasan (termasuk PPh dipotong pembeli) tidak melebihi sisa."""
    target = max(0.0, float(sisa_pelunasan or 0))
    if target <= 0:
        return 0.0
    if str(getattr(kontrak, "is_pph", "false") if kontrak else "false").lower() != "true":
        return round(target)

    lo, hi = 0, int(target)
    best = 0
    while lo <= hi:
        mid = (lo + hi) // 2
        if effective_pelunasan(mid, None, kontrak) <= target + 0.5:
            best = mid
            lo = mid + 1
        else:
            hi = mid - 1
    return float(best)